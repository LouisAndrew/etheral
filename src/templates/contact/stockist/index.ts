
import { Stockist } from './stockist';

export default Stockist;