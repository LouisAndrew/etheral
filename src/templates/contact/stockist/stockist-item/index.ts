import { StockistItem, Props } from './stockist-item';

export { Props };

export default StockistItem;
