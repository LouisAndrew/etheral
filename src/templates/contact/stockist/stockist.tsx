import React from 'react';
import { graphql, useStaticQuery } from 'gatsby';
import { groupBy, toPairs } from 'lodash';

import { Box, Heading } from 'rebass';
import { InlineIcon } from '@iconify/react';
import mapPin2Fill from '@iconify/icons-ri/map-pin-2-fill';

import StockistItem, { Props as StockistItemProps } from './stockist-item';

type Props = {};

// how to render stockist => use lodash groupBy() => group and render stockist(s) by location.
// more or less the same thing as consignment.
const Stockist: React.FC<Props> = () => {
    const data = useStaticQuery(graphql`
        query {
            allConsignment {
                edges {
                    node {
                        address
                        location
                        name
                        phoneNumber
                        web
                        zipCode
                    }
                }
            }
        }
    `);

    if (data) {
        // apply upercase to every consignemnt's location => case safety on grouping
        const consignmentsQuery = data.allConsignment.edges
            .map((edge: any) => edge.node)
            .map((csg: any) => ({
                ...csg,
                location: csg.location.toUpperCase(),
            }));

        // then grouping them by location.
        const consignments = groupBy(consignmentsQuery, 'location');

        return (
            <Box ml={[0, 0, 6, 8]} mt={[5, 5]}>
                <Heading
                    as="h4"
                    variant="h4"
                    textAlign={['center', 'center', 'left']}
                >
                    STOCKISTS
                </Heading>
                <Box mt={[4, 4]}>
                    {/* to pairs => map object key and value into an array. [0] = key, [1] = value */}
                    {toPairs(consignments).map(pair => (
                        <Box data-testid={pair[0]} as="ul" key={pair[0]}>
                            <Heading
                                variant="h4"
                                sx={{
                                    display: 'flex',
                                    alignItems: 'center',
                                    justifyContent: [
                                        'center',
                                        'center',
                                        'flex-start',
                                    ],
                                    svg: { mr: [] },
                                }}
                            >
                                <InlineIcon icon={mapPin2Fill} />
                                {pair[0]}
                            </Heading>
                            <>
                                {pair[1].map(val => (
                                    <StockistItem
                                        key={val.name}
                                        {...(val as StockistItemProps)}
                                    />
                                ))}
                            </>
                        </Box>
                    ))}
                </Box>
            </Box>
        );
    } else {
        return null;
    }
};

export { Stockist };
