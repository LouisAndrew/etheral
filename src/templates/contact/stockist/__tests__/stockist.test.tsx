import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import { render, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';

import { Stockist } from '../stockist';

import * as Gatsby from 'gatsby';

const stockistJktA = {
    address: '',
    location: 'Jakarta',
    name: 'Toko A',
    phoneNumber: '123',
    web: '',
    zipCode: '123123',
};

const stockistJktB = {
    address: '',
    location: 'Jakarta',
    name: 'Toko B',
    phoneNumber: '1123',
    web: '',
    zipCode: '12',
};

const stockistBdg = {
    address: '',
    location: 'Bandung',
    name: 'Toko C',
    phoneNumber: '1123',
    web: '',
    zipCode: '12',
};

describe('Stockist', () => {
    const useStaticQuery = jest.spyOn(Gatsby, 'useStaticQuery');

    const Element = <Stockist />;

    beforeEach(() => {
        useStaticQuery.mockImplementationOnce(() => ({
            allConsignment: {
                edges: [
                    { node: stockistJktA },
                    { node: stockistJktB },
                    { node: stockistBdg },
                ],
            },
        }));
    });
    afterEach(cleanup);

    it('renders without crashing', () => {
        const div = document.createElement('div');
        ReactDOM.render(Element, div);
    });

    it('Should render based on location correctly', () => {
        const { getByRole, getByTestId } = render(Element);

        const Jkt = getByTestId('JAKARTA');
        const Bdg = getByTestId('BANDUNG');

        // check for listitems on Jkt
        const LiJkt = document.querySelectorAll(`${Jkt.className} li`);
        LiJkt.forEach(el => {
            expect(
                el.textContent?.includes('Toko A') ||
                    el.textContent?.includes('Toko B')
            ).toBe(true);
        });

        const LiBdg = document.querySelectorAll(`${Bdg.className} li`);
        LiBdg.forEach(el => {
            expect(el).toHaveTextContent('Toko C');
        });
    });

    it('matches snapshot', () => {
        const run = false;

        if (run) {
            const tree = renderer.create(Element).toJSON();
            expect(tree).toMatchSnapshot();
        }
    });
});
