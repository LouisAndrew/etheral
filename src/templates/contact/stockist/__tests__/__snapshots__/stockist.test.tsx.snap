// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stockist matches snapshot 1`] = `
<div
  className="css-15eedg5"
>
  <h4
    className="css-2xeirz"
  >
    STOCKISTS
  </h4>
  <div
    className="css-1ynh1v9"
  >
    <ul
      className="css-vurnku"
      data-testid="JAKARTA"
    >
      <h2
        className="css-12osdm3"
      >
        <svg
          dangerouslySetInnerHTML={
            Object {
              "__html": "<path d=\\"M18.364 17.364L12 23.728l-6.364-6.364a9 9 0 1 1 12.728 0zM12 13a2 2 0 1 0 0-4a2 2 0 0 0 0 4z\\" fill=\\"currentColor\\"/>",
            }
          }
          focusable={false}
          height="1em"
          preserveAspectRatio="xMidYMid meet"
          style={
            Object {
              "transform": "rotate(360deg)",
              "verticalAlign": "-0.125em",
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        />
        JAKARTA
      </h2>
      <li
        className="css-1t6fcji"
      >
        <div
          className="css-m65dgr"
        >
          Toko A
        </div>
        <p
          className="css-1li2d9"
        >
          
          <br />
          123123
          <br />
          123
          <br />
        </p>
        <a
          className="css-abcknd"
          href="http://"
        >
          
        </a>
      </li>
      <li
        className="css-1t6fcji"
      >
        <div
          className="css-m65dgr"
        >
          Toko B
        </div>
        <p
          className="css-1li2d9"
        >
          
          <br />
          12
          <br />
          1123
          <br />
        </p>
        <a
          className="css-abcknd"
          href="http://"
        >
          
        </a>
      </li>
    </ul>
    <ul
      className="css-vurnku"
      data-testid="BANDUNG"
    >
      <h2
        className="css-12osdm3"
      >
        <svg
          dangerouslySetInnerHTML={
            Object {
              "__html": "<path d=\\"M18.364 17.364L12 23.728l-6.364-6.364a9 9 0 1 1 12.728 0zM12 13a2 2 0 1 0 0-4a2 2 0 0 0 0 4z\\" fill=\\"currentColor\\"/>",
            }
          }
          focusable={false}
          height="1em"
          preserveAspectRatio="xMidYMid meet"
          style={
            Object {
              "transform": "rotate(360deg)",
              "verticalAlign": "-0.125em",
            }
          }
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        />
        BANDUNG
      </h2>
      <li
        className="css-1t6fcji"
      >
        <div
          className="css-m65dgr"
        >
          Toko C
        </div>
        <p
          className="css-1li2d9"
        >
          
          <br />
          12
          <br />
          1123
          <br />
        </p>
        <a
          className="css-abcknd"
          href="http://"
        >
          
        </a>
      </li>
    </ul>
  </div>
</div>
`;
